### DISCUSSION TITLE
Service Discovery

### DISCUSSION THREAD

**Discussion by hemalbavishi** [eclipse-iceoryx/iceoryx#1167] [Q&A]
Hello,

I am facing an issue with Iceoryx when I try to integrate with ara::com. I have three different process. 

**### Scenario 1:**
First process is sending data over SOME/IP, second process is consuming data from first process over SOME/IP IPC and sending same data to third process via Iceoryx IPC. 

Configuration:
Process 1: Service(Publisher) : "Video"
                 Instance : "FrontRight"
                 Event : "VideoData"
                 Communication : "SOME/IP-IPC"
Process 2: Service(Subscriber) : "Video"
                 Instance : "FrontRight"
                 Event : "VideoData"
                 Communication : "SOME/IP-IPC"
                Service(Publisher) : "Image"
                 Instance : "FrontRight"
                 Event : "ImageData"
                 Communication : "Iceoryx-IPC"
Process 3: Service(Subscriber) : "Image"
                 Instance : "FrontRight"
                 Event : "ImageData"
                 Communication : "Iceoryx-IPC"

Observation:
I am able to offer service from second process and third process and third process is also discovering service which I offered from the second process. Third process is able to discover service offered by second process. Moreover, communication is also established between second and third process i.e. I am able to send data from the second process and same data is received by the third process.
------------------------------------------------------------------------------------------------------------------------------------------------
**### Scenario 2:**
In order to measure performance, I configured loop back mode between second process and third process. It means after receiving data from the second process, third process is sending it back to second process to measure end to end performance for the same service but with different instance. 

Configuration:
Process 1: Service(Publisher) : "Video"
                 Instance : "FrontRight"
                 Event : "VideoData"
                 Communication : "SOME/IP-IPC"
Process 2: Service(Subscriber) : "Video"
                 Instance : "FrontRight"
                 Event : "ImageData"
                 Communication : "SOME/IP-IPC"
                Service(Publisher) : "Image"
                 Instance : "FrontRight"
                 Event : "ImageData"
                 Communication : "Iceoryx-IPC"
                 Service(Subscriber) : "Image"
                 Instance : "FrontLeft"
                 Event : "ImageData"
                 Communication : "Iceoryx-IPC"
Process 3: Service(Subscriber) : "Image"
                 Instance : "FrontRight"
                 Event : "VideoData"
                 Communication : "Iceoryx-IPC"  
                Service(Subscriber) : "Image"
                 Instance : "FrontLeft"
                 Event : "VideoData"
                 Communication : "Iceoryx-IPC"

**Observation: 
I am able to offer service from second process and third process and vice versa. Third process is not able to discover service which I offered from the second process but second process is able to discover service which I offered from third process. This problem is observed only in Service Discovery but event communication is happening as expected.** 
------------------------------------------------------------------------------------------------------------------------------------------------
Any clue in this regard is highly appreciated.

**Comment by MatthiasKillat** [comment]
@hemalbavishi  Could you provide the `findService` calls (with the respective inputs) and their results. Then we can discuss which of them behave as expected and which do not. Especially whether you are using wildcards etc.

If communication is established it in the system any app (sending app, receiving app, any other app) should be able to find the service by using the `findService` API.

**Reply by hemalbavishi** [reply]
Hello Matthias,

Thank you!
**### Publisher**
`auto testService = iox::capro::ServiceDescription{iox::capro::IdString_t{iox::cxx::TruncateToCapacity,  iox::cxx::convert::toString(service_id)}, iox::capro::IdString_t{iox::cxx::TruncateToCapacity,  iox::cxx::convert::toString(instance_id)}, iox::capro::IdString_t{}};
publishers_.emplace_back(testService);
auto& publisher = publishers_.back();
publisher.offer();`

**### Subscriber**
`if(!callbackAlreadyRegistered) {
        std::cout << "Registring Callback Handler <"  << std::endl;
        autosar_discovery.registerCallback(std::bind(&IceoryxProxyFactoryImpl::RegisterAvailabilityHandler, this, std::placeholders::_1));
    }`

`auto result = discovery.findService(iox::capro::IdString_t{iox::cxx::TruncateToCapacity,  iox::cxx::convert::toString(handle.serviceId)}, iox::capro::IdString_t{iox::cxx::TruncateToCapacity,  iox::cxx::convert::toString(*optionalId)}, iox::capro::Wildcard);
    if (result.size() > 0)
    {
        std::cout << "Discovery Monitor <" << handle.serviceId << ", " << handle.instanceId.ToString() <<  " > available"
                    << std::endl;
        serviceInfo.callback({iceoryx_mapping->GetProxy(*optionalId)}, handle);            
    }
    else
    {
       std::cout << "Discovery Monitor <" << handle.serviceId << ", " << handle.instanceId.ToString() << ", " <<  "> unavailable"
                    << std::endl;
       serviceInfo.callback({}, handle);
    }`

**Comment by hemalbavishi** [comment]
Hello Matthias,

Looks like issue is not with Iceoryx but ara::com binding. I fixed it and now I am able to establish communication between Process 2 and Process 3 or vice versa. I still have some teething issue but it seems to be more in ara::com and not in iceoryx:) Nevertheless,  my next step is to measure performance with ara::com binding. I will keep you posted once I have result with me.   

Thank you so much for your help.
